demo/ota-agent/ota_demo.c

```diff
@@ -0,0 +1,409 @@
+{
+    uint32_t ulIndex = 0;
+    OtaDataEvent_t * freeBuffer = NULL;
+
+    if( xSemaphoreTake( bufferSemaphore, portMAX_DELAY ) == pdTRUE )
+    {
+        for( ulIndex = 0; ulIndex < MAX_NUM_OF_OTA_DATA_BUFFERS; ulIndex++ )
+        {
+            if( dataBuffers[ ulIndex ].bufferUsed == false )
+            {
+                dataBuffers[ ulIndex ].bufferUsed = true;
+                freeBuffer = &dataBuffers[ ulIndex ];
+                break;
+            }
+        }
+
+        ( void ) xSemaphoreGive( bufferSemaphore );
+    }
+    else
+    {
+        printf("Failed to get buffer semaphore. \n" );
+    }
+
+    return freeBuffer;
+}
+
+void otaDemo_start( void )
+{
+    OtaEventMsg_t initEvent = { 0 };
+
+    if( !mqttWrapper_isConnected() )
+    {
+        return;
+    }
+
+    bufferSemaphore = xSemaphoreCreateMutex();
+
+    if( bufferSemaphore != NULL )
+    {
+        memset( dataBuffers, 0x00, sizeof( dataBuffers ) );
+    }
+
+    OtaInitEvent_FreeRTOS();
+
+    initEvent.eventId = OtaAgentEventRequestJobDocument;
+    OtaSendEvent_FreeRTOS( &initEvent );
+
+    while (otaAgentState != OtaAgentStateStopped ) {
+        processOTAEvents();
+    }
+}
+
+OtaState_t getOtaAgentState()
+{
+    return otaAgentState;
+}
+
+static void requestJobDocumentHandler()
+{
+    char thingName[ MAX_THING_NAME_SIZE + 1 ] = { 0 };
+    size_t thingNameLength = 0U;
+
+    mqttWrapper_getThingName( thingName, &thingNameLength );
+    coreJobs_startNextPendingJob( thingName,
+                                  strnlen( thingName, MAX_THING_NAME_SIZE ),
+                                  "test",
+                                   4U );
+}
+
+static void initMqttDownloader( AfrOtaJobDocumentFields_t *jobFields )
+{
+    char thingName[ MAX_THING_NAME_SIZE + 1 ] = { 0 };
+    size_t thingNameLength = 0U;
+
+    numOfBlocksRemaining = jobFields->fileSize /
+                            mqttFileDownloader_CONFIG_BLOCK_SIZE;
+    numOfBlocksRemaining += ( jobFields->fileSize %
+                            mqttFileDownloader_CONFIG_BLOCK_SIZE > 0 ) ? 1 : 0;
+    currentFileId = jobFields->fileId;
+    currentBlockOffset = 0;
+    totalBytesReceived = 0;
+
+    mqttWrapper_getThingName( thingName, &thingNameLength );
+
+    /* Initialize the File downloader */
+    mqttDownloader_init( &mqttFileDownloaderContext,
+                        jobFields->imageRef,
+                        jobFields->imageRefLen,
+                        thingName,
+                        thingNameLength,
+                        DATA_TYPE_JSON );
+}
+
+static bool receivedJobDocumentHandler( OtaJobEventData_t * jobDoc )
+{
+    bool parseJobDocument = false;
+    bool handled = false;
+    char * jobId;
+    size_t jobIdLength = 0U;
+    AfrOtaJobDocumentFields_t jobFields = { 0 };
+
+    jobIdLength = coreJobs_getJobId( (char *)jobDoc->jobData, jobDoc->jobDataLength, &jobId );
+
+    /* TODO: Change length to the length of global job id */
```

Comment:
Does this still need to be done?

Reply:
yes, I think we need to revisit this part to avoid the scenario where the jobIDLength is greater than the maximum job id length.